source/tutorials/nixos-configuration-testing.md

```diff
@@ -0,0 +1,485 @@
+```
+
+you need to remove the symbolic link and only than you can remove the cached result:
+```shell-session
+$ rm ./result
+```
+
+now you are able to remove the cached result in the nix store directly
+```shell-session
+$ nix-store --delete /nix/store/4klj06bsilkqkn6h2sia8dcsi72wbcfl-vm-test-run-unnamed
+```
+
+### wayland application example (todo remove home-manager)
+
+The configuration we are using is starting the gnome desktop manager using wayland.
+To test if a wayland application is working is more complicated because we need to automate the login into gnome and automated startup of the application. Additionally we need to enable access to gnome dbus interface. To do this we need to modify the configuration
+
+ the automated start of the application including automated login to gnome/wayland
+
+
+In the machine configuration we need to enable autologin for the user alice.
+
+```nix
+      services.xserver.displayManager.autoLogin.enable = true;
+      services.xserver.displayManager.autoLogin.user = "alice";
+```
+
+To simplify our script we pin the uid of the user to 1000.
+```nix
+        uid = 1000;
+```
+
+We specify a service that auto start firefox after login, which is easier than doing this in the test script.
+
+```nix
+      environment.systemPackages = [
+        (pkgs.makeAutostartItem {
+          name = "firefox";
+          package = pkgs.firefox;
+        })
+      ];
+```
+<!-- check again if this is best practice -->
+
+
+Because gnome doesn't allow the evaluation of javascript to get information about open windows we need to override the gnome-shell startup service to start gnome-shell in unsafe mode:
+
+```nix
+      systemd.user.services = {
+        "org.gnome.Shell@wayland" = {
+          serviceConfig = {
+            ExecStart = [
+              # Clear the list before overriding it.
+              ""
+              # Eval API is now internal so Shell needs to run in unsafe mode.
+              "${pkgs.gnome.gnome-shell}/bin/gnome-shell --unsafe-mode"
+            ];
+          };
+        };
+```
+
+The test script utilizes the gnome dbus interface to get a list of open wayland windows. we wait until firefox appear to be started and make a screenshot that will be found in the result folder.
+
+```{code-block}
+    testScript = {nodes, ...}: let
+      user = nodes.machine.config.users.users.alice;
+      bus = "DBUS_SESSION_BUS_ADDRESS=unix:path=/run/user/${toString user.uid}/bus";
+      gdbus = "${bus} gdbus";
+      su = command: "su - ${user.name} -c '${command}'";
+      gseval = "call --session -d org.gnome.Shell -o /org/gnome/Shell -m org.gnome.Shell.Eval";
+      wmClass = su "${gdbus} ${gseval} global.display.focus_window.wm_class";
+    in ''
+      machine.wait_until_succeeds("${wmClass} | grep -q 'firefox'")
+      machine.sleep(20)
+      machine.screenshot("screen")
+    '';
+```
```

Comment:
An optional explanation (hidden in `<details><summary>Detailed explanation</summary> ... </details>`) of exactly *what the hell is happening* in this script would be an order.